src/bin/pg_autoctl/keeper.c

```diff
@@ -2190,3 +2213,176 @@ keeper_call_reload_hooks(Keeper *keeper, bool firstLoop)
 	/* we're done reloading now. */
 	asked_to_reload = 0;
 }
+
+
+/*
+ * keeper_read_nodes_from_file read the keeper->config.pathnames.nodes file (a
+ * JSON Array of Nodes with id, name, host, port, lsn, and is_primary) and
+ * fills in the internal keeper otherNodes array. Use that function when the
+ * monitor is disabled.
+ */
+bool
+keeper_read_nodes_from_file(Keeper *keeper, NodeAddressArray *nodesArray)
+{
+	KeeperConfig *config = &(keeper->config);
+	KeeperStateData *state = &(keeper->state);
+
+	char *contents = NULL;
+	long size = 0L;
+
+	/* refrain from reading the nodes list when in the INIT state */
+	if (state->current_role == INIT_STATE)
+	{
+		return true;
+	}
+
+	/* if the file does not exists, we're done */
+	if (!file_exists(config->pathnames.nodes))
+	{
+		log_debug("Nodes files \"%s\" does not exists, done processing",
+				  config->pathnames.nodes);
+		return true;
+	}
+
+	if (!read_file_if_exists(config->pathnames.nodes, &contents, &size))
+	{
+		log_error("Failed to read nodes array from file \"%s\"",
+				  config->pathnames.nodes);
+		return false;
+	}
+
+	/* now parse the nodes JSON file */
+	if (!parseNodesArray(contents, nodesArray, state->current_node_id))
+	{
+		log_debug("Failed to parse JSON nodes array:\n%s", contents);
+		log_error("Failed to parse nodes array from file \"%s\"",
+				  config->pathnames.nodes);
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * keeper_get_primary fetches the current primary Node in the group, either by
+ * connecting to the monitor and using the pgautofailover.get_primary() API
+ * there, or by scanning through the keeper->otherNodes array for the first
+ * node with isPrimary true.
+ *
+ * In both cases, there might not be a primary node identified at the moment,
+ * in which case we return false.
+ */
+bool
+keeper_get_primary(Keeper *keeper, NodeAddress *primaryNode)
+{
+	KeeperConfig *config = &(keeper->config);
+
+	if (!config->monitorDisabled)
+	{
+		Monitor *monitor = &(keeper->monitor);
+
+		if (!monitor_get_primary(monitor,
+								 config->formation,
+								 keeper->state.current_group,
+								 primaryNode))
+		{
+			log_error("Failed to get the primary node from the monitor, "
+					  "see above for details");
+			return false;
+		}
+
+		return true;
+	}
+	else
+	{
+		for (int i = 0; i < keeper->otherNodes.count; i++)
+		{
+			NodeAddress *node = &(keeper->otherNodes.nodes[i]);
+
+			if (node->isPrimary)
```

Comment:
Should we also check for group here? Or is nodes.json only for a single group?

Reply:
Yeah the nodes.json file is intended for a single group.